### PR TITLE
Support for SIR intrinsics (closes #1826)

### LINKED ISSUE #1826: Update SIR to use Dawn intrinsics (min, max etc).
At the moment Fortran intrinsics are replaced by equivalent code before being transformed to SIR. However, Dawn does support a number of intrinsics so these can be translated directly.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
The generated PSyIR has been checked by Matthias and he is happy that it is correct.

**Comment by rupertford** [comment]
Ready for first review from @arporter or @sergisiso

**Comment by rupertford** [comment]
Note, the output has been checked by Matthas @ meteoswiss.

File: examples/nemo/eg4/tra_adv_compute.F90 [hunk 1]
```diff
@@ -4,51 +4,33 @@ module tra_adv_compute_mod
...
-                zice(ji,jj,jk)                               &
+             IF( tsn(ji,jj,jk) <= ztfreez(ji,jj) + 0.1d0 ) THEN   ;   zice = 1.d0
+             ELSE                                                 ;   zice = 0.d0
+             ENDIF
+             zind(ji,jj,jk) = MAX (                &
+                rnfmsk(ji,jj) * rnfmsk_z(jk),      &
+                upsmsk(ji,jj)               ,      &
+                zice                               &
                 &                  ) * tmask(ji,jj,jk)
-             zind(ji,jj,jk) = 1 - zind(ji,jj,jk)
```
Comment (arporter) [inline examples/nemo/eg4/tra_adv_compute.F90:51]: How come this line has been removed?

File: src/psyclone/psyir/backend/sir.py [hunk 2]
```diff
@@ -317,6 +314,10 @@ def binaryoperation_node(self, node):
         operator to SIR.
 
         '''
+        intrinsic_operators = [
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py:317]: Please add a comment saying what this list is for and again for `binary_operators` below.

File: src/psyclone/psyir/backend/sir.py [hunk 3]
```diff
@@ -520,3 +552,39 @@ def schedule_node(self, node):
...
+        the PSyIR tree.
+
+        :param node: a NaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.NaryOperation`
+
+        :returns: the SIR Python code.
+        :rtype: str
+
+        :raises VisitorError: if there is no mapping from the PSyIR \
+        operator to SIR.
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py:567]: Please indent this  continuation by 4 spaces.

File: src/psyclone/psyir/backend/sir.py [hunk 4]
```diff
@@ -520,3 +552,39 @@ def schedule_node(self, node):
...
+
+        '''
+        nary_operators = {
```
Comment (arporter) [inline src/psyclone/psyir/backend/sir.py:570]: Please add a comment giving the purpose of this dict - I guess this relates to the comment at L587?

File: src/psyclone/tests/psyir/backend/sir_test.py [hunk 5]
```diff
@@ -942,4 +940,83 @@ def test_sirwriter_schedule_node_1(parser, sir_writer):
...
+            f"make_literal_access_expr(\"1.0\", BuiltinType.Float)], "
+            f"[make_literal_access_expr(\"2.0\", BuiltinType.Float)])"
+            in result)
+
+
+def test_sirwriter_binaryoperation_sign_node(parser, sir_writer):
+    '''Check the binaryoperation_node method of the SIRWriter class
+    outputs the expected SIR code when the binaryoperation is sign
+    intrinsic. This is a special case as the sign intrinsic is
+    implemented differently in Fortran and C.
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sir_test.py:978]: Picky but really we're only talking about PSyIR and SIR here?

File: src/psyclone/tests/psyir/backend/sir_test.py [hunk 6]
```diff
@@ -942,4 +940,83 @@ def test_sirwriter_schedule_node_1(parser, sir_writer):
...
+    unsupported operator.
+
+    '''
+    # nary SUM is not currently supported in the SIR backend.
+    code = CODE.replace("1.0", "SUM(1.0, 2.0, 3.0)")
+    rhs = get_rhs(parser, code)
+    with pytest.raises(VisitorError) as info:
+        _ = sir_writer.naryoperation_node(rhs)
+    assert ("Method naryoperation_node in class SIRWriter, unsupported "
+            "operator 'Operator.SUM' found." in str(info.value))
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/sir_test.py:1003]: Would be good if this error message said which operators *are* supported too.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good - just some minor things to tidy.
Changes are pylint clean.
Tests and examples are all fine with compilation.
Updated documentation builds fine.
I have no way of testing that the generated SIR is correct. Please could you show me some output to verify that it works for the tracer advection example.

File: doc/user_guide/transformations.rst [hunk 7]
```diff
@@ -1036,9 +1036,16 @@ SIR
...
+for the NEMO API. The first `sir_trans.py` simply outputs SIR. This
+will raise an exception if used with the `tracer advection` example as
+the example contains array-index notation which is not supported by
+the SIR backend, but will generate code for the other examples. The
+second, `sir_trans_loop.py` includes transformations to hoist code out
+of a loop, translate array-index notation into explicit loops and
+translate a single access to an array dimension to a one-trip loop (to
+make the code suitable for the SIR backend). This works with the
+`tracer-advection` example. The third script `sir_trans_all.py`
+additionally additionally replaces any intrinsics with equivalent code
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/transformations.rst:1055]: Oops!

File: examples/nemo/eg4/sir_trans_loop.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+from psyclone.psyir.nodes import Assignment
+from psyclone.psyir.transformations import HoistTrans
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans, \
+    NemoAllArrayAccess2LoopTrans
+
+
+def trans(psy):
+    '''Transformation routine for use with PSyclone. Applies the PSyIR2SIR
+    transform to the supplied invokes. This transformation is limited
```
[the remark this answers is not in the document]
Comment (rupertford) [reply examples/nemo/eg4/sir_trans_loop.py:54]: Oops.

File: examples/nemo/eg4/sir_trans_loop.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+from psyclone.psyir.backend.sir import SIRWriter
+
+from psyclone.psyir.nodes import Assignment
+from psyclone.psyir.transformations import HoistTrans
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans, \
+    NemoAllArrayAccess2LoopTrans
+
+
+def trans(psy):
+    '''Transformation routine for use with PSyclone. Applies the PSyIR2SIR
```
[the remark this answers is not in the document]
Comment (rupertford) [reply examples/nemo/eg4/sir_trans_loop.py:53]: Oops. Copy and paste error I expect.

File: examples/nemo/eg4/sir_trans_loop.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        # (tra_adv_compute.F90).
+        for loop in schedule.loops():
+            # outermost only
+            if loop.loop_type == "levels":
+                for child in loop.loop_body[:]:
+                    if isinstance(child, Assignment):
+                        hoist_trans.apply(child)
+
+        kern = sir_writer(schedule)
+        print(kern)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply examples/nemo/eg4/sir_trans_loop.py:99]: Added todo and a new issue #1854

File: examples/nemo/eg4/tra_adv_compute.F90 (hunk 1 above)
Comment (rupertford) [reply examples/nemo/eg4/tra_adv_compute.F90:51]: I don't know. Adding it back in does not affect code generation so I've reinstated it.

File: src/psyclone/psyir/backend/sir.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py:317]: Added comment.

File: src/psyclone/psyir/backend/sir.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py:567]: Oops.

File: src/psyclone/psyir/backend/sir.py (hunk 4 above)
Comment (rupertford) [reply src/psyclone/psyir/backend/sir.py:570]: Added.

File: src/psyclone/tests/psyir/backend/sir_test.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/sir_test.py:978]: Modified.

File: src/psyclone/tests/psyir/backend/sir_test.py (hunk 6 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/sir_test.py:1003]: OK, added.

**Comment by rupertford** [comment]
Ready for next review from @arporter. The validation of the code generation was done by Matthias. He took the output and checked it was OK. Did you want me to try to run the SIR with Dawn locally as well?

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Rupert assures me that Matthias says it works :-)
All tests and examples are OK with compilation.
Ref. guide builds with no new warnings. UG builds fine.
Will proceed to merge.

**Pull request merged**